core/src/main/scala/org/apache/spark/ui/jobs/StagePage.scala

```diff
@@ -434,25 +470,208 @@ private[ui] class StagePage(parent: StagesTab) extends WebUIPage("stage") {
       val content =
         summary ++
         showAdditionalMetrics ++
+        makeTimeline(stageData.taskData.values.toSeq, currentTime) ++
         <h4>Summary Metrics for {numCompleted} Completed Tasks</h4> ++
         <div>{summaryTable.getOrElse("No tasks have reported metrics yet.")}</div> ++
         <h4>Aggregated Metrics by Executor</h4> ++ executorTable.toNodeSeq ++
         maybeAccumulableTable ++
         <h4>Tasks</h4> ++ taskTable
-
       UIUtils.headerSparkPage("Details for Stage %d".format(stageId), content, parent)
     }
   }
 
+  def makeTimeline(tasks: Seq[TaskUIData], currentTime: Long): Seq[Node] = {
```

Comment:
Can we make it so that we have some task count `MAX_TIMELINE_TASKS` and if there are more tasks than that we don't generate and send this data? The issue is that currently this UI has scalability issues because the scala XML output can create very large amounts of intermediate data and actually cause the driver to OOM. For now, maybe if there are more than 1000 or so tasks we should not include this. Over time we can increase the count if we find that it is okay.